docs/architecture-and-design.md

```diff
@@ -0,0 +1,388 @@
+![Coordinate Convention Diagram](images/angle_convention_diagram.svg)
+
+Because of our [Coordinate Conventions](#coordinates), this means that an angle of `0` will always face the enemy net not matter which side of the field we are defending in reality.
+
+# Architecture Overview
+At a high-level our system is made of 3 main components: The [Backend](#backend), the [AI](#ai), and the [Visualizer](#visualizer). These 3 components each run in their own thread, and communicate with each other using the [Observer design pattern](#observer-design-pattern). Together, they are what make up our AI.
+
+The Backend is responsible for communicating with the outside world (network and radio), the AI is what makes the actual gameplay decisions, and the Visualizer shows us what's happening and lets us control the AI.
+
+Each component is described in more detail in their own sections.
+
+
+#### Architecture Overview Diagram
+![High Level Architecture Diagram](images/high_level_architecture_diagram.svg)
+
+
+# Backend
+The `Backend` is responsible for all communication with the "outside world". The responsabilities of the `Backend` can be broken down into Input and Output.
+
+### Input Responsabilities
+1. Receiving robot status messages
+2. Receiving vision data about where the robots and ball are (typically provided by [SSL-Vision](#ssl-vision) or [grSim](#grsim))
+2. Receiving referee commands (typically from the [SSL-Gamecontroller](#ssl-gamecontroller)
+3. Filtering the received data
+    * **Why we need to do this:** Programs that provide data like [SSL-Vision](#ssl-vision) only provide raw data. This means that if there are several orange blobs on the field, [SSL-Vision](#ssl-vision) will tell us the ball is in several different locations. It is up to us to filter this data to determine the "correct" position of the ball. The same idea applies to robot positions and other data we receive.
+4. Storing the filtered data into the [World](#world) datastructures understood by our system
+5. Sending the filtered data to the rest of the system
+
+### Output Responsabilities
+1. Sending robot primitives to the robots
+
+In practice, the `Backend` is just a simple interface that specifies [World](#world) and [Robot Status](#robot-status) objects must be produced, and [Primitves](#primitives) may be consumed. The interface is very generic so that different implementations may be swapped out in order to communicate with different hardware / protocols / programs. For example, we have multiple implementations of the "output" part of the backend: one that lets us send data to our real robots using the radio, and one that sends commands to simulated robots in [grSim](#grsim).
```

Comment:
"must be produced, and [Primitives]*(#primitives)"

Reply:
fixed